src/mame/dataeast/dec8.h

```diff
@@ -17,216 +17,299 @@
-	TILE_GET_INFO_MEMBER(get_ghostb_fix_tile_info);
-	TILE_GET_INFO_MEMBER(get_oscar_fix_tile_info);
+	void allocate_buffered_spriteram16();
+	void dec8_s_map(address_map &map);
+	void oscar_s_map(address_map &map);
+
+};
+
+// with I8751 MCU
+class mcu_state : public base_state
+{
+public:
+	mcu_state(const machine_config &mconfig, device_type type, const char *tag) :
+		base_state(mconfig, type, tag),
+		m_mcu(*this, "mcu"),
+		m_coin_port(*this, "I8751")
+	{
+	}
+protected:
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+
+	TIMER_CALLBACK_MEMBER(mcu_irq_clear);
+
+	required_device<i8751_device> m_mcu;
+
+	/* ports */
+	required_ioport m_coin_port;
+
+	u8 i8751_h_r();
+	u8 i8751_l_r();
+	void i8751_w(offs_t offset, u8 data);
+
+	u8 i8751_port0_r();
+	void i8751_port0_w(u8 data);
+	u8 i8751_port1_r();
+	void i8751_port1_w(u8 data);
+
+	void i8751_reset_w(u8 data);
+
+	// MCU communication
+	u8  m_i8751_p2 = 0;
+	int m_i8751_port0 = 0;
+	int m_i8751_port1 = 0;
+	int m_i8751_return = 0;
+	int m_i8751_value = 0;
+
+	emu_timer *m_i8751_timer = nullptr;
+};
+
+// with unique sprite hardware
+class srdarwin_state : public mcu_state
+{
+public:
+	srdarwin_state(const machine_config &mconfig, device_type type, const char *tag) :
+		mcu_state(mconfig, type, tag)
+	{
+	}
+	void srdarwin(machine_config &config);
+
+protected:
+	virtual void video_start() override;
+
+private:
+	void mcu_to_main_w(u8 data);
+	void srdarwin_videoram_w(offs_t offset, u8 data);
+	void control_w(offs_t offset, u8 data);
+
+	TILE_GET_INFO_MEMBER(get_fix_tile_info);
+	TILE_GET_INFO_MEMBER(get_tile_info);
+
+	uint32_t screen_update(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+	void draw_sprites(bitmap_ind16 &bitmap, const rectangle &cliprect, bitmap_ind8 &primap);
+
+	void main_map(address_map &map);
+};
+
+// with 'karnov' sprite hardware
+class lastmisn_state : public mcu_state
+{
+public:
+	lastmisn_state(const machine_config &mconfig, device_type type, const char *tag) :
+		mcu_state(mconfig, type, tag),
+		m_spritegen_krn(*this, "spritegen_krn"),
+		m_nmigate(*this, "nmigate")
+	{
+	}
+
+	void garyoret(machine_config &config);
+	void ghostb(machine_config &config);
+	void lastmisn(machine_config &config);
+	void meikyuh(machine_config &config);
+	void shackled(machine_config &config);
+
+	void init_meikyuhbl();
+
+protected:
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+
+	required_device<deco_karnovsprites_device> m_spritegen_krn;
+	optional_device<input_merger_device> m_nmigate;
+
+	void ghostb_bank_w(u8 data);
+	void gondo_scroll_w(offs_t offset, u8 data);
+	void gondo_mcu_to_main_w(u8 data);
+
+	TILE_GET_INFO_MEMBER(get_gondo_fix_tile_info);
+	TILE_GET_INFO_MEMBER(get_gondo_tile_info);
+
+	DECLARE_VIDEO_START(lastmisn);
+	uint32_t screen_update_lastmisn(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+
+	void screen_vblank(int state);
```

Comment:
Please keep the data members grouped and member functions grouped within the `protected` section – you have the `m_spritegen_krn` and `m_nmigate` intermixed with the member functions.